<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Document</title>
</head>

<body>
  <script>
    /* 
      test函数有两个参数，这两个参数都是函数，如果执行成功会调用resolve()，如果执行失败会调用reject()
      可以看出，test()只关心自己的逻辑，并不关心具体的resolve和reject将如何处理结果
     */
    function test(resolve, reject) {
      var timeOut = Math.random() * 2;
      console.log(`set timeout to ${timeOut} seconds`);
      setTimeout(function () {
        if (timeOut < 1) {
          console.log("call resolve()...");
          resolve("200 ok");
        } else {
          console.log("call reject()...");
          reject(`timeout in ${timeOut} seconds...`);
        }
      }, timeOut * 1000)
    }

    /* 
    有了执行函数，可以用一个Promise对象来执行它，并在将来的某个时候获得成功或失败的结果 
    p1 是一个Promise对象，它负责执行test函数。test函数内部是异步执行的
    */
    var p1 = new Promise(test);
    var p2 = p1.then(function (result) {
      console.log(`成功: ${result}`);
    })
    var p3 = p2.catch(function (reason) {
      console.log(`失败: ${reason}`);
    })

    /* Promise对象可以串联起来，所以可以简化为： */
    new Promise(test).then(function (result) {
      console.log(`成功：${result}`);
    }).catch(function (reason) {
      console.log(`失败：${reason}`);
    });
  </script>
</body>

</html>